x-pack/plugin/security/lib/jose-wrapper/src/main/java/org/elasticsearch/nimbus/NimbusWrapper.java

```diff
@@ -0,0 +1,106 @@
+import org.elasticsearch.SpecialPermission;
+
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+import java.security.PrivilegedActionException;
+import java.security.PrivilegedExceptionAction;
+import java.text.ParseException;
+import java.util.Map;
+
+/**
+ * This class wraps the operations requiring access in {@link AccessController#doPrivileged(PrivilegedAction)} blocks.
+ * Can't do these operations inline with giving too much access due to how the security manager calculates the stack for lambda expressions.
+ * Isolating the calls here allows for least privilege access to this helper jar.
+ */
+public class NimbusWrapper {
+
+    // utility class
+    private NimbusWrapper() {}
+
+    public static String getHeaderAsString(SignedJWT signedJWT) {
+        SpecialPermission.check();
+        return AccessController.doPrivileged((PrivilegedAction<String>) () -> signedJWT.getHeader().toString());
+
+    }
+
+    public static String getClaimsSetAsString(JWTClaimsSet jwtClaimsSet) {
+        SpecialPermission.check();
+        return AccessController.doPrivileged((PrivilegedAction<String>) jwtClaimsSet::toString);
+    }
+
+    public static JWTClaimsSet verifyTokenClaims(IDTokenValidator validator, JWT idToken, Nonce nonce) throws BadJOSEException,
+        JOSEException {
+        try {
+            return AccessController.doPrivileged(
+                (PrivilegedExceptionAction<JWTClaimsSet>) () -> validator.validate(idToken, nonce).toJWTClaimsSet()
+            );
+        } catch (PrivilegedActionException exception) {
+            if (exception.getCause() instanceof BadJOSEException e) {
+                throw e;
+            } else if (exception.getCause() instanceof JOSEException e) {
+                throw e;
+            } else {
+                throw new ElasticsearchException(exception);
+            }
+        }
+    }
+
+    // only used in tests
+    public static SignedJWT newSignedJwt(JWSHeader header, JWTClaimsSet claimsSet) {
+        SpecialPermission.check();
+        return AccessController.doPrivileged((PrivilegedAction<SignedJWT>) () -> new SignedJWT(header, claimsSet));
+    }
+
+    // only used in tests
+    public static SignedJWT newSignedJWT(Map<String, Object> header, JWTClaimsSet claimsSet, String signatureUrl) throws ParseException {
+        SpecialPermission.check();
+        try {
+            return AccessController.doPrivileged(
+                (PrivilegedExceptionAction<SignedJWT>) () -> new SignedJWT(
+                    JWSHeader.parse(header).toBase64URL(),
+                    claimsSet.toPayload().toBase64URL(),
+                    Base64URL.encode(signatureUrl)
+                )
+            );
+        } catch (PrivilegedActionException e) {
+            if (e.getException() instanceof ParseException ex) {
+                throw ex;
+            } else {
+                throw new RuntimeException(e.getException());
```

Comment:
We should probably use ElasticsearchException since it can serialize itself, even if we never expect this to happen. (same below).